### ISSUE TITLE
Bug: "does not have the requested state 'takeoff'"

### ISSUE THREAD

**Issue by legoboyvdlp** [c172p-team/c172p#1271]
Every so often when I start on the runway with the launcher I get this error message:

![image](https://user-images.githubusercontent.com/10434911/56794785-d25e5200-6806-11e9-8e77-f257eaf132b7.png)

I just left my state at the "default state" option.

I thought we added state support a while back though?

**Comment by gilbertohasnofb** [comment]
@legoboyvdlp is this still a bug?

**Comment by legoboyvdlp** [comment]
Yes, still seeing this almost every startup.

**Comment by gilbertohasnofb** [comment]
@wlbragg can you reproduce this? Is this being caused by the initial states system?

**Comment by wlbragg** [comment]
You must be throwing a parameter  --state="takoff" or something to that effect?
We never finished the state work. I had a fairly elaborate state branch at one point but scrapped it and pushed only the bare minimum that supported starting in the air. Part of the issue was the state stuff and the persistent stuff doesn't play nicely with each other. Timing of when all this stuff loads was an issue. To do this right we really need to completely rethink the entire aircraft data system we have setup and plan for that and the state stuff to work together.
I'd have to review the files to provide any more information.

**Comment by wkitty42** [comment]
i think, but not totally sure, that the start states and persistent state conflicts need to be handled in core... i say this because i'm not aware of when/where aircraft-data is loaded into the sim...

my initial thoughts say that the persistent should be loaded first and then the selected start state overlaid on that but that could mean that the start states must take care to switch things (eg tie downs, cowling covers, etc) as needed for being in that selected state...

either one overlays the other and then flips things as needed or if using a start state, persistent data is not loaded but then you lose your (eg) seat positioning and similar...

**Comment by wlbragg** [comment]
I can't really say it is a core issue. What I think is needed in our case is a way to switch saved data loading on and off, which is core I guess but it's kind of a fix after the fact. Then when using a state mode you can throw the "do not load  persistent data" switch. The problem was that the persistent data loaded after the state stuff and the only way around it was to use a local version of persistent data where we would save what we want to load ourselves VS the built in aircraft data which is defeating the whole purpose of the built in persistent data functions.
I had come to the conclusion if the persistent data system is left as-is, as in there is no "don't load data" switch, we needed to refactor the current persistent data and move all of its functionality to states. Our built in routine we have developed along the away "save current status of the aircraft" is the main issue. The problem is it saves the data every time so the exiting state of the aircraft is always available at startup. We purposely overwrite that data every time the "save state of the aircraft" is checked. If the State stuff loaded second, then it would fit perfectly with our prior system of retaining the exiting state of the aircraft. But it doesn't, it loads first. That was at the heart of the issue.

**Comment by legoboyvdlp** [comment]
No, I never pass any explicit state. This just happens more or less on every start. My state setting has always been at the default.

**Comment by wlbragg** [comment]
For you to get that message a "state" request the aircraft doesn't have has to be passed to the launcher or FG (I'm not sure where the alert is generated). So if your not passing it as a command line switch then you are selecting a state in the launcher that is not available and the launcher is sending that request for a state that doesn't exist. Maybe you have something stuck in the aircraft data, but I don't think that request ever gets saved as far as I know.
Something is different in your system than everyone else then, because no one else is getting that. I just tried to set the launcher startup to any available states, there are only two, default and automatic, neither throw a switch or request that causes that message. What version are you running? Can you inspect any arguments that are being passed?

**Comment by legoboyvdlp** [comment]
But I'm explicitly not setting a state -- I'm leaving everything as it is out of the box? 😕

**Comment by wlbragg** [comment]
>So if your not passing it as a command line switch then you are selecting a state in the launcher that is not available

I should have said the launcher may be passing a request based on what you have chosen in the location tab. But that seems unlikely because it's not happening to anyone else.  Any way you can video tape your entire startup sequence?

**Comment by legoboyvdlp** [comment]
Yes, so it is, as shown by Ctrl-L.

Looking in LauncherController.cxx, we see that 

`if (m_selectedAircraftInfo->hasState("take-off")) {
        return "take-off";
    }`

This occurs if:
`if ((m_aircraftState == "auto") && !m_selectedAircraftInfo->haveExplicitAutoState()) {`

Which occurs if 

` if (m_selectedAircraftInfo->hasStates() && !m_aircraftState.isEmpty()) {`

The question is, why this occurs here and not for others. I also don't see it every time -- it appears almost as if it occurs the second launch only? 

But the problem is that I specifically left it at "default", certainly not "automatic". So, somehow m_aircraftState is becoming "auto" if it isn't explicitly set. 


I notice that if I explicitly click on "default" in the state selection window, everything is as normal.
